wpilib/wpilib/iterativerobotbase.py

```diff
@@ -0,0 +1,226 @@
+# validated: 2017-11-09 TW ef3267833fc3 edu/wpi/first/wpilibj/IterativeRobotBase.java
+# ----------------------------------------------------------------------------
+# Copyright (c) FIRST 2008-2012. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+# ----------------------------------------------------------------------------
+
+import hal
+import logging
+
+from .robotbase import RobotBase
+from .livewindow import LiveWindow
+
+__all__ = ["IterativeRobotBase"]
+
+
+class IterativeRobotBase(RobotBase):
+    """IterativeRobotBase implements a specific type of robot program framework, extending the RobotBase
+    class.
+
+    The IterativeRobotBase class does not implement startCompetition(), so it should not be used
+    by teams directly.
+
+    This class provides the following functions which are called by the main loop,
+    startCompetition(), at the appropriate times:
+
+    robotInit() -- provide for initialization at robot power-on
+
+    init() functions -- each of the following functions is called once when the
+    appropriate mode is entered:
+    - disabledInit()   -- called only when first disabled
+    - autonomousInit() -- called each and every time autonomous is entered from
+    another mode
+    - teleopInit()     -- called each and every time teleop is entered from
+    another mode
+    - testInit()       -- called each and every time test is entered from
+    another mode
+
+    periodic() functions -- each of these functions is called on an interval:
+    - robotPeriodic()
+    - disabledPeriodic()
+    - autonomousPeriodic()
+    - teleopPeriodic()
+    - testPeriodic()
+    """
+
+    class Mode:
+        kNone = 0
+        kDisabled = 1
+        kAutonomous = 2
+        kTeleop = 3
+        kTest = 4
+
+    last_mode = Mode.kNone
+
+    #: A python logging object that you can use to send messages to the log. It
+    #: is recommended to use this instead of print statements.
+    logger = logging.getLogger("robot")
+
+    def __init__(self):
+        super().__init__()
+        hal.observeUserProgramStarting()
+
+    def startCompetition(self):
+        raise NotImplementedError("You haven't implemented startCompetition")
```

Comment:
This isn't needed here, right?

Reply:
Yes it is?

Reply:
Why? Just let it go through the inheritance chain.

Reply:
Oh you mean let RobotBase give the error?

Reply:
Hm. I think it's there for documentation sake. This being here has _negative_ utility if it's not documented.

Reply:
Yeah idk why they implemented it just to keep it abstract...If I get rid of it here, I wanna move the message because `RobotBase` doesn't give anything

Reply:
The message is tautological, realistically. It's in the traceback.

Reply:
Eh true